Georgena/hybrid cache topology

georgen117 opened the pull request:
## Description

This PR fixes inaccurate per-core cache size reporting on hybrid x86 CPUs
(Alder Lake, Raptor Lake, Meteor Lake, Lunar Lake, Nova Lake) by adding an
x64-specific platform layer that uses `Xbyak::util::CpuTopology` instead of
CPUID leaf 4.

### Problem

The legacy `get_per_core_cache_size(int level)` uses CPUID leaf 4 via
`Xbyak::util::Cpu`. On hybrid systems the CPUID sharing-count fields are buggy
or ambiguous, producing values that do not match what any core type actually
sees. Measured examples:

| Platform | Cache | Legacy (CPUID) | New (topology) | Issue |
|----------|-------|----------------|----------------|-------|
| ARL | L2 | 512 KB | 1024 KB (lp_core) | ~6x CPUID undercount |
| PTL | L2 | 384 KB | 1024 KB (lp_core) | ~8x CPUID undercount |
| PTL | L1d | 24 KB | 32 KB (lp_core) | CPUID halves L1d even with no HT |
| LNL | L2 | 320 KB | 1024 KB (lp_core) | ~8x CPUID undercount |
| NVL | L2 | 512 KB | 1024 KB (lp_core) | ~4x CPUID undercount |
| MTL | L2 | 512 KB | 2048 KB (p_core) | ~4x CPUID undercount |

<img width="1069" height="597" alt="image" src="https://github.com/user-attachments/assets/dd1d0869-05ec-4d1e-8a90-3c25056f33cd" />

A hard-coded `jit_brgemm_conv_utils.cpp` L2 override of 2 MB for AVX2 was
masking part of this on those platforms. Changing the hard-coded L2 override addresses
[MFDNN-14078](https://jira.devtools.intel.com/browse/MFDNN-14078)

### Solution

A new x64-specific module (`src/cpu/x64/platform.cpp/.hpp`) uses
`Xbyak::util::CpuTopology` and distinguishes three core classes:

| Core class | Description |
|------------|-------------|
| P-core     | Intel Performance core |
| LP-core    | Efficient core sharing an L3 cache |
| LPE-core   | Efficient core with no L3 (e.g. Meteor Lake SoC tile) |

On hybrid systems the implementation now always returns a conservative cache
budget:

- `min` policy in code: `min(p_core, lp_core[, lpe_core])` per cache level.
- no runtime policy override.
- no policy enum exposed through the x64 platform API.

`min` is the default because over-committing cache is more costly than
under-utilising it: a blocking heuristic sized for P-cores would overflow the
smaller E-core caches when the workload runs there. Empirical comparison during
development (`p_core` vs `min` on NVL/PTL P-core-pinned runs) showed mixed
per-shape results with no stable rule, confirming `min` as the safer general
default.

<!-- optional: insert "At a glance" screenshot from
     percore_policy_presentation.xlsx showing the mixed NVL result -->

### Performance validation (A/B, hardcode vs min)

Convolution data was collected to validate that changing the hardcoded L2 cache size to
a size based on detected cache is the right behavior. A/B data collection was used to
help remove thermal and run-order issues, i.e. multiple runs in different call orders
to avoid spikes caused by thermal throttling.

Ratio = hardcode / min
- If ratio > 1.0, hardcode took more time than min, so min is faster.
- If ratio < 1.0, hardcode took less time than min, so hardcode is faster.
- If ratio ~= 1.0, they are effectively neutral.


<img width="586" height="222" alt="image" src="https://github.com/user-attachments/assets/9e38a514-c28d-4f86-be55-ab3e56896115" />

The MFDNN-14078 shape in standalone testing saw runtime increase from 41.7 ms to
98.6 ms (about 2.36x). However, with A/B testing, the MFDNN-14078 result shrank to
only slightly better than hardcoded performance (about 1.01x).

**Note:** data was collected with `--cold-cache=all` data was recollected without the `--cold-cache`
this resulted in small differences in performance but no outliers that would significantly change the
conclusion based on the `--cold-cache=all` results.

**Performance impact:** first call on a hybrid system costs ~25-30 ms to
initialise `CpuTopology` (magic static, paid once). On non-hybrid systems this
cost is bypassed entirely -- the non-hybrid fast path returns the CPUID value
immediately.

The default `get_per_core_cache_size(int level)` in `src/cpu/platform.cpp`
delegates to `x64::platform::get_per_core_cache_size(level)` on x64 and falls
back to the CPUID path on all other architectures.

All cache debuginfo logging for x64 now lives in `src/cpu/x64/platform.cpp`
(`print_cache_debuginfo_once` and `print_hybrid_cache_debuginfo_once`) behind
`ONEDNN_VERBOSE=debuginfo=1`.

### Files changed

| File | Change |
|------|--------|
| `src/cpu/x64/platform.hpp` | New x64 cache-topology API declaration for `get_per_core_cache_size(int level)` and hybrid helpers. |
| `src/cpu/x64/platform.cpp` | Implementation using `Xbyak::util::CpuTopology`. macOS guard (falls back to legacy). Thread-safe lazy init via C++11 magic statics. Non-hybrid fast path avoids `CpuTopology` init. Includes x64 cache debuginfo printing. |
| `src/cpu/platform.cpp` | Delegates `get_per_core_cache_size(int level)` to `x64::platform::get_per_core_cache_size(level)` on x64; AARCH64 path preserved. |
| `src/cpu/x64/jit_brgemm_conv_utils.cpp` | Removes hard-coded AVX2 L2 = 2 MB override. L2 is now always `get_per_core_cache_size(2) * 3/4`. |

### Known limitations

- `Xbyak::util::CpuTopology` is not available on macOS; all macOS paths fall
  back to legacy CPUID behaviour.
- All-E-core systems (e.g. Sierra Forest) are treated as non-hybrid.

dzarukin commented on src/cpu/platform.cpp:
```diff
@@ -350,6 +351,839 @@ size_t get_timestamp() {
+
+    if (!global_cache_topology.is_hybrid) {
+        // For Non-hybrid system, all cores are assumed to be p-core
+        return pcore_size;
```
What about SRF where all cores are e-cores?

georgen117 replied:
Just to make sure I did test the code on a SRF system and it behaved just like a non-hybrid system which is what I expected.

dzarukin commented on src/cpu/platform.cpp:
```diff
@@ -350,6 +351,839 @@ size_t get_timestamp() {
+    //       other paths associated with cache information.
+
+    // Read CPU information from /sys/devices/system/cpu/
+    for (int cpu = 0; cpu < 256; cpu++) { // Check up to 256 CPUs
```
Isn't this number quarriable?

georgen117 commented on src/cpu/platform.cpp:
```diff
@@ -350,6 +351,839 @@ size_t get_timestamp() {
 #endif
 }
 
+bool is_hybrid() {
```
Thanks for the suggestion. I placed the code in `cpu/platform.cpp` because I was not really clear where to place the files. Moving it into `cpu/x64/platform.cpp` makes since.

georgen117 commented on src/cpu/platform.cpp:
```diff
@@ -350,6 +351,839 @@ size_t get_timestamp() {
+
+//------------------Start CPUID cache topology code-----------------------------
+
+void populate_cache_map_cpuid(
```
The functions are guarded by std::once found in the `init_cache_topology()` function. However, there is nothing in each of the OS specific functions that will ensure thread safety.

georgen117 committed:
cpu: x64: fix per-core cache size on hybrid systems

The legacy get_per_core_cache_size() uses CPUID leaf 4 via
Xbyak::util::Cpu. On hybrid systems the CPUID cache sharing-count
is inaccurate, causing cache size, particularly L2 cache, to be
significantly under reported on hybrid systems.

A hard-coded L2 override of 2 MB for AVX2 in
jit_brgemm_conv_utils.cpp had masked part of the issue on those
platforms.

On hybrid systems get_per_core_cache_size() now returns a conservative
min(p_core, lp_core[, lpe_core]) per cache level.

Non-hybrid systems continue to use the legacy CPUID path directly,
bypassing CpuTopology init (25-30 ms).

Removes the hard-coded AVX2 L2 = 2 MB override in
jit_brgemm_conv_utils.cpp. L2 is now always get_per_core_cache_size(2).

Cache topology is logged once under ONEDNN_VERBOSE=debuginfo=1.

Adds src/cpu/x64/platform.{hpp,cpp} using Xbyak::util:CpuTopology

georgen117 commented on src/cpu/platform.cpp:
```diff
@@ -350,6 +351,839 @@ size_t get_timestamp() {
+    init_cache_topology_linux(global_cache_topology);
+#else
+    // Fallback: use CPUID-based method
+    init_cache_topology_cpuid(global_cache_topology);
```
The comment is on old code  I did take the time to measure the time:

| Path | First-call cost | Subsequent calls |
|---|---|---|
| New `CpuTopology` init | ~25ms - 30ms | negligible (cached static) |
| Legacy (CPUID) | ~7µs - 10µs (existing `xbyak::Cpu` already cached) | negligible (existing `xbyak::Cpu` already cached) |

Due to this measurement, I did a small reorganization of the code if the system is not hybrid, then the code will automatically default to the legacy CPUID path.  This avoids the cost of initializing the `CpuTopology` Unfortantly this cost is unavoidable if we want to get accurate cache size on hybrid systems and the cost is one time on init.

densamoilov commented on src/common/verbose.cpp:
```diff
@@ -109,6 +109,47 @@ void print_header() noexcept {
         verbose_printf("info,cpu,isa:%s\n", cpu::platform::get_isa_info());
+        if (cpu::platform::is_hybrid_cpu()) {
+            verbose_printf(
+                    "info,cpu,per_core_cache,pcore:L1d:%uKB,L2:%uKB,L3:%uKB\n",
```
I'm not sure how we expect this information to be used. If we already know what CPU is used we can find this information.
I'm also not sure how useful it would be to have this information in verbose in general. Given that you have the most experience working with it, any use cases?

georgen117 replied:
This was really useful when validating this PR. However, I don't know if there is a wider use for this information. Removing the verbose logging would simplify the PR.

georgen117 replied:
I removed all the verbose_printf out of the verbose.cpp function and moved it into the cpu/x64/platform.cpp under debuginfo logging.

Since a log of the functions placed in cpu/platform.hpp were only placed there to make them avalible to the verbose print I was able to remove all those functions reducing the diff from this PR considerably.

densamoilov commented on src/common/verbose.cpp:
```diff
@@ -109,6 +109,47 @@ void print_header() noexcept {
+        // policy. Emit a cache_budget line showing the effective values so a
+        // reader of the log can see exactly what the blocking heuristics used.
+        {
+            const std::string policy = getenv_string_user("CACHE_POLICY");
```
We need to document environment variables. 

I actually think that it'd probably be better if we put it under `debuginfo`, assuming this info is generally useful.

georgen117 replied:
I will remove the environment variable. The environment variable was added to validate the code making it possible to change the behavior without recompiling the code.

Along with the removal of the verbose print statments the PR will be further simplified.

georgen117 replied:
I removed the CACHE_POLICY environment variable.

densamoilov commented on src/cpu/x64/jit_brgemm_conv_utils.cpp:
```diff
@@ -1655,13 +1655,8 @@ status_t init_jcp(jit_brgemm_conv_conf_t &jcp, cpu_isa_t isa,
-        brg_blocking_t::L2 = 2 * 1024 * 1024;
     // take L2 as 3/4 of the real size for L2
-    brg_blocking_t::L2 = (brg_blocking_t::L2 * 3) / 4;
+    brg_blocking_t::L2 = (platform::get_per_core_cache_size(2) * 3) / 4;
```
The PR doesn't seem to have perf data. I remember the problem was that this change affected performance for some cases. Do you have perf data to verify that it doesn't cause any regressions?

georgen117 replied:
I know that this fixes the performance reported for  [MFDNN-14078](https://jira.devtools.intel.com/browse/MFDNN-14078).

I will need to do a quick check of the data I collected to give you more concrete numbers for other shapes. I may need to collect some number with the older hardcoded L2 value.

georgen117 replied:
see the **Performance validation (A/B, hardcode vs min)** section in the PR description for numbers.  I used A/B testing for the performance gathering. The  difference between the hardcoded value and the min value was minimal. I tried to collect the same shape data as was collected for PR#2420 which introduced the hardcoded value. I did the best I could with the data found in `tests/benchdnn/inputs/conv/`

densamoilov commented on src/cpu/x64/platform.cpp:
```diff
@@ -0,0 +1,236 @@
+    // Validate level
+    if (level < 1 || level > 3) { return 0; }
+
+#ifdef __APPLE__
```
I'm not sure we need to handle the APPLE case for x64 platforms

@vpirogov, do we?

vpirogov replied:
Apple dropped support for x64 in macOS 29. While macOS 28 has 3 more years of life I don't think we need to care for anything other than breaking the build.

densamoilov commented on src/cpu/x64/platform.hpp:
```diff
@@ -0,0 +1,82 @@
+// On non-hybrid CPUs, the cache_sizing_policy_t argument is ignored and the function
+// returns the per-core cache size as normal.
+//
+// - if cache_sizing_policy_t is p_core/lp_core/lpe_core, the function returns the per-core cache
```
What is the rationale behind having the sizing policies? Why do we need this level of flexibility?

georgen117 replied:
The original thought was that some shapes that were compute bound instead of memory bound would be able to choose the p_core cache_sizing_policy_t and get an increase in performance. I looked for a heuristic that could benefit from changing to a specific cache_size_policy_t and I couldn't come up with one.  I can remove the cache_sizing_policy_t all together and have the code default to the `min` value for hybrid cores and default the legacy (CPUID leaf 4) behavior for non-hybrid systems.

That would further simplify the code. It could always be re-introduced if I can successfully find a case that the performance can be improved by using the `cache_sizing_policy_t`

georgen117 replied:
when I removed the `CACHE_POLICY` environment variable I also removed the `cache_sizing_policy_t` env variable since the environment variable was the only way to change the policy.

georgen117 reviewed (commented):
@densamoilov after reading your reviews I can simplify this PR considerably.  This was over complicated because I was investigating a lot of possible configureations.  I ended up on using CPUID leaf 4 for non-hybrid systems and the `per_core_cache_size(min(p_core, lp_core[, lpe_core])` if I remove the envronment varable the is only two paths that are decided by the existence of the hybrid CPUID flag. 

I will simplify the PR.
